File: 1401-1500/1461-1470/1466M_Reorder_Routes_to_Make_All_Paths_Lead_to_the_City_Zero/cs/bfs.cs
// 1466. Reorder Routes to Make All Paths Lead to the City Zero
// https://leetcode.com/problems/reorder-routes-to-make-all-paths-lead-to-the-city-zero/
// T.C.: O(n)
// S.C.: O(n)
public class Solution
{
  public int MinReorder(int n, int[][] connections)
  {
    var changes = 0;
    var roads = new List<int>[n];
    for (var i = 0; i < n; i++)
    {
      roads[i] = new List<int>();
    }

    foreach (var connection in connections)
    {
      var a = connection[0];
      var b = connection[1];
      roads[a].Add(b);
      roads[b].Add(-a);
    }

    var visited = new bool[n];
    var queue = new Queue<int>();
    queue.Enqueue(0);
    visited[0] = true;

    while (queue.Count > 0)
    {
      var city = queue.Dequeue();
      foreach (var nextCity in roads[city])
      {
        if (visited[Math.Abs(nextCity)]) continue;

        if (nextCity > 0) changes++;
        visited[Math.Abs(nextCity)] = true;
        queue.Enqueue(Math.Abs(nextCity));
      }
    }

    return changes;
  }
}

var n = 6;
var connections = new int[][] {
  new int[] { 0, 1 },
  new int[] { 1, 3 },
  new int[] { 2, 3 },
  new int[] { 4, 0 },
  new int[] { 4, 5 },
 };
var expected = 3;
var result = new Solution().MinReorder(n, connections);
Console.WriteLine($"{result}, {result == expected}");

n = 5;
connections = new int[][] {
  new int[] { 1, 0 },
  new int[] { 1, 2 },
  new int[] { 3, 2 },
  new int[] { 3, 4 },
};
expected = 2;
result = new Solution().MinReorder(n, connections);
Console.WriteLine($"{result}, {result == expected}");

n = 3;
connections = new int[][] {
  new int[] { 1, 0 },
  new int[] { 2, 0 },
};
expected = 0;
result = new Solution().MinReorder(n, connections);
Console.WriteLine($"{result}, {result == expected}");

File: 1401-1500/1461-1470/1466M_Reorder_Routes_to_Make_All_Paths_Lead_to_the_City_Zero/bfs.js
// 1466. Reorder Routes to Make All Paths Lead to the City Zero
// https://leetcode.com/problems/reorder-routes-to-make-all-paths-lead-to-the-city-zero/
// T.C.: O(n)
// S.C.: O(n)
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var minReorder = function (n, connections) {
  let changes = 0;
  const roads = Array.from({ length: n }, () => []);
  for (const [a, b] of connections) {
    roads[a].push(b);
    roads[b].push(-a);
  }

  const visited = new Array(n).fill(false);
  const queue = [0];
  visited[0] = true;

  while (queue.length) {
    const city = queue.shift();
    if (roads[city].length === 0) continue;
    for (const nextCity of roads[city]) {
      if (visited[Math.abs(nextCity)]) continue;

      if (nextCity > 0) changes++;
      visited[Math.abs(nextCity)] = true;
      queue.push(Math.abs(nextCity));
    }
  }

  return changes;
};

var n = 6,
  connections = [
    [0, 1],
    [1, 3],
    [2, 3],
    [4, 0],
    [4, 5],
  ];
var expected = 3;
var result = minReorder(n, connections);
console.log(result, result === expected);

var n = 5,
  connections = [
    [1, 0],
    [1, 2],
    [3, 2],
    [3, 4],
  ];
var expected = 2;
var result = minReorder(n, connections);
console.log(result, result === expected);

var n = 3,
  connections = [
    [1, 0],
    [2, 0],
  ];
var expected = 0;
var result = minReorder(n, connections);
console.log(result, result === expected);
